### PR TITLE
New checker `unnecessary-list-index-lookup` (#4525)

### LINKED ISSUE #4525: New checker ``unnecessary-list-index-lookup``
As discussed originally in PR #4508 

```python
for i, _ in enumerate(self.options_providers):
    if provider.priority > self.options_providers[i].priority: #  the index lookup here is unnecessary
        self.options_providers.insert(i, provider)
        break
```

### PR DESCRIPTION
**Pull request by timmartin**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
|    | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

Added a new checker for catching the case of iterating over a list using `enumerate()`:

    for index, value in enumerate(mylist):
        print(mylist[index]) # Bad, should use value instead

This will now report

    Unnecessary list index lookup, use 'value' instead

Closes #4525

### CODE REVIEW

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Thank you for opening the pull request, this look goods already. I think we'll be able to add this to 2.14, but we need to release 2.13 at some point and we're lagging behind all the contributions 😄. Could you move the changelog to 2.14 ? I know the "what's new" page for 2.14 is not created yet, sorry.

File: tests/functional/u/unnecessary/unnecessary_list_index_lookup.py
Comment (timmartin) [reply tests/functional/u/unnecessary/unnecessary_list_index_lookup.py:44]: I actually already tried that one when I first wrote the cases, but it turns out to be a syntax error:

    SyntaxError: assignment expression cannot rebind comprehension iteration variable 'val'

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (timmartin) [reply pylint/checkers/refactoring/refactoring_checker.py]: Good point, I've done this.

File: pylint/checkers/__init__.py
Comment (DanielNoord) [inline pylint/checkers/__init__.py:126]: @timmartin Could you maybe extract these changes unrelated to adding the checker itself (but actually fixing issues) in a separate PR? We're likely to squash merge this PR (we generally do so) so the commit message for this change would be a bit misleading.

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/refactoring_checker.py:1561]: Can you add the message to the decorator here?

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/refactoring_checker.py:1979]: Correct me if I am wrong, but wouldn't this crash for:
```python
for i in enumerate(list()):
    ...
```

I don't think we can assume `elts` has length `2`. In normal Python code we should, but `pylint` shouldn't crash when people or working on their code and lint an intermediate phase.

Could you test this and add a test case for this?

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/refactoring_checker.py:2028]: Could you add a `confidence` here? It think it should be `HIGH` as there is no inference (but perhaps double check this).

File: tests/functional/u/unnecessary/unnecessary_list_index_lookup.py
Comment (DanielNoord) [inline tests/functional/u/unnecessary/unnecessary_list_index_lookup.py:1]:
```suggestion
"""Tests for unnecessary-list-index-lookup."""
# pylint: disable=missing-docstring, too-few-public-methods, expression-not-assigned, line-too-long, unused-variable
```

File: tests/functional/u/unnecessary/unnecessary_list_index_lookup.py
Comment (DanielNoord) [reply tests/functional/u/unnecessary/unnecessary_list_index_lookup.py:40]: That sounds like a good opportunity for a new checker :)

File: tests/functional/u/unnecessary/unnecessary_list_index_lookup_py38.py
Comment (DanielNoord) [inline tests/functional/u/unnecessary/unnecessary_list_index_lookup_py38.py:1]:
```suggestion
"""Tests for unnecessary-list-index-lookup with assignment expressions."""
# pylint: disable=missing-docstring, too-few-public-methods, expression-not-assigned, line-too-long, unused-variable
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]
This looks very good already!

I left some comments, let me know if you have any questions about them 😄

File: pylint/checkers/__init__.py
Comment (timmartin) [reply pylint/checkers/__init__.py:126]: Done. I'll leave these changes in this branch for now, or it will fail in CI. Will this get rebased before the merge so that we can eliminate these lines from the squashed commit?

Comment (DanielNoord) [reply pylint/checkers/__init__.py:126]: Depends on the timeline of this PR I think. Pierre scheduled it for `2.14` since we really want to devote all our attention to getting `2.13` asap, but I had some spare time and went ahead and reviewed already. Depending on the time when we really get to reviewing this and the state of `main` we might need to rebase or merge `main` itself.

Comment (DanielNoord) [reply pylint/checkers/__init__.py:126]: I merged the changes so you could merge `main` to eliminate the lines from this PR/commits.

Comment (timmartin) [reply pylint/checkers/__init__.py]: Done. I did a squash while I was at it.

File: tests/functional/u/unnecessary/unnecessary_list_index_lookup.py
Comment (timmartin) [reply tests/functional/u/unnecessary/unnecessary_list_index_lookup.py:46]: I'll leave that to someone else! Since it's a syntax error, it won't be going unnoticed anyway.

File: pylint/checkers/__init__.py
Comment (timmartin) [reply pylint/checkers/__init__.py]: Tests are all passing now. Let me know if there are any other things for me to do here.

File: pylint/checkers/refactoring/refactoring_checker.py
Comment (DanielNoord) [inline pylint/checkers/refactoring/refactoring_checker.py:1987]: I'm wondering if there are edge cases to consider with `len(args) > 1`, but I think in those cases using the index is so unnecessarily complicated that most people will have moved to using `value` at that point. So I think we can leave those edge cases unconsidered here.

**Review by DanielNoord — APPROVED** [review]
Thanks a lot @timmartin, this looks good to me!

@Pierre-Sassoulas Depending on your schedule and free time, we might actually get this reviewed and merged before we release `2.11`. No rush though!

**Comment by DanielNoord** [comment]
Hi @timmartin,

I know I already approved this PR, but...
We have been working on the documentation of the messages that pylint emits. As this PR adds a new message could I ask you to rebase on `main` and add two examples for bad and good code like explained in https://github.com/PyCQA/pylint/issues/5953? That way we immediately help users understand what this new message will do 😄

**Comment by timmartin** [comment]
I've rebased and added documentation now.

**Review by Pierre-Sassoulas — APPROVED** [review]
Thank you for the two examples, looks great !

**Pull request merged**
